File: bratek20-architecture/src/test/java/pl/bratek20/architecture/events/api/EventPublisherTest.java
package pl.bratek20.architecture.events.api;

import lombok.Value;
import org.junit.jupiter.api.Test;
import pl.bratek20.architecture.context.spring.SpringContextBuilder;
import pl.bratek20.architecture.events.impl.EventsContextModule;
import pl.bratek20.tests.InterfaceTest;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class EventPublisherTest extends InterfaceTest<EventPublisher>{

    @Value
    public static class TestEvent implements Event {
        int value;
    }

    @Value
    public static class OtherEvent implements Event {
        int value;
    }

    @Value
    public class TestEventListener implements EventListener<TestEvent> {
        List<TestEvent> events = new ArrayList<>();

        @Override
        public void handleEvent(TestEvent event) {
            events.add(event);
        }
    }

    @Value
    public class OtherEventListener implements EventListener<OtherEvent> {
        List<OtherEvent> events = new ArrayList<>();

        @Override
        public void handleEvent(OtherEvent event) {
            events.add(event);
        }
    }

    private TestEventListener listener1;
    private TestEventListener listener2;
    private OtherEventListener otherListener;

    @Override
    protected EventPublisher createInstance() {
        listener1 = new TestEventListener();
        listener2 = new TestEventListener();
        otherListener = new OtherEventListener();

        return new SpringContextBuilder()
            .setObject(listener1)
            .setObject(listener2)
            .setObject(otherListener)
            .withModule(new EventsContextModule())
            .build()
            .get(EventPublisher.class);
    }

    @Test
    void shouldWork() {
        // given

        // when
        instance.publish(new TestEvent(1));
        instance.publish(new TestEvent(2));
        instance.publish(new OtherEvent(3));

        // then
        assertThat(listener1.events)
            .containsExactly(
                new TestEvent(1),
                new TestEvent(2)
            );
        assertThat(listener2.events)
            .containsExactly(
                new TestEvent(1),
                new TestEvent(2)
            );
        assertThat(otherListener.events)
            .containsExactly(
                new OtherEvent(3)
            );
    }
}

File: bratek20-architecture/src/test/java/pl/bratek20/architecture/events/api/EventPublisherTest.kt
package pl.bratek20.architecture.events.api

import org.assertj.core.api.Assertions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import pl.bratek20.architecture.context.api.ContextBuilder
import pl.bratek20.architecture.context.guice.GuiceContextBuilder
import pl.bratek20.architecture.context.spring.SpringContextBuilder
import pl.bratek20.architecture.events.impl.EventsContextModule
import java.util.stream.Stream

class EventPublisherTest {
    data class TestEvent(
        val value: Int
    ) : Event


    data class OtherEvent(
        val value: Int
    ) : Event

    class TestEventListener : EventListener<TestEvent> {
        var events: MutableList<TestEvent> = ArrayList()

        override fun handleEvent(event: TestEvent) {
            events.add(event)
        }
    }

    class OtherEventListener : EventListener<OtherEvent> {
        var events: MutableList<OtherEvent> = ArrayList()

        override fun handleEvent(event: OtherEvent) {
            events.add(event)
        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("provideArguments")
    fun shouldWork(name: String?, builder: ContextBuilder) {
        // given
        val c = builder
            .addImpl(EventListener::class.java, TestEventListener::class.java)
            .addImpl(EventListener::class.java, OtherEventListener::class.java)
            .withModule(EventsContextModule())
            .build()

        val listeners = c.getMany(EventListener::class.java)

        val listener = listeners.find { l: EventListener<*> -> l is TestEventListener } as TestEventListener
        val otherListener = listeners.find { l: EventListener<*> -> l is OtherEventListener } as OtherEventListener

        val publisher = c.get(EventPublisher::class.java)

        // when
        publisher.publish(TestEvent(1))
        publisher.publish(TestEvent(2))
        publisher.publish(OtherEvent(3))

        // then
        Assertions.assertThat(listener.events)
            .containsExactly(
                TestEvent(1),
                TestEvent(2)
            )
        Assertions.assertThat(otherListener.events)
            .containsExactly(
                OtherEvent(3)
            )
    }

    companion object {
        @JvmStatic
        fun provideArguments(): Stream<Arguments> {
            return Stream.of(
                Arguments.of("Spring", SpringContextBuilder()),
                Arguments.of("Guice", GuiceContextBuilder())
            )
        }
    }
}